File: .taira-source-c80bf4d85a97/java/iroha_android/src/main/java/org/hyperledger/iroha/android/crypto/MlDsaSigner.java
package org.hyperledger.iroha.android.crypto;

import org.hyperledger.iroha.android.SigningException;

/** ML-DSA signer backed by the native {@code connect_norito_bridge} helpers. */
public final class MlDsaSigner implements Signer {
  private final MlDsaPrivateKey privateKey;
  private final MlDsaPublicKey publicKey;

  public MlDsaSigner(final MlDsaPrivateKey privateKey) {
    this(privateKey, privateKey.publicKey());
  }

  public MlDsaSigner(final MlDsaPrivateKey privateKey, final MlDsaPublicKey publicKey) {
    if (privateKey == null) {
      throw new IllegalArgumentException("privateKey must not be null");
    }
    if (publicKey == null) {
      throw new IllegalArgumentException("publicKey must not be null");
    }
    this.privateKey = privateKey;
    this.publicKey = publicKey;
  }

  @Override
  public byte[] sign(final byte[] message) throws SigningException {
    if (message == null) {
      throw new SigningException("message must not be null");
    }
    try {
      final byte[] prehashed = IrohaHash.prehash(message);
      return NativeSignerBridge.signDetached(
          SigningAlgorithm.ML_DSA, privateKey.getEncoded(), prehashed);
    } catch (final RuntimeException ex) {
      throw new SigningException("ML-DSA signing failed", ex);
    }
  }

  @Override
  public byte[] publicKey() {
    return publicKey.getEncoded();
  }

  @Override
  public String algorithm() {
    return SigningAlgorithm.ML_DSA.providerName();
  }
}

File: .taira-source-c80bf4d85a97/kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/crypto/MlDsaSigner.kt
package org.hyperledger.iroha.sdk.crypto

/** ML-DSA signer backed by the native `connect_norito_bridge` helpers. */
class MlDsaSigner(
    private val privateKey: MlDsaPrivateKey,
    private val keyPublicKey: MlDsaPublicKey = privateKey.publicKey(),
) : Signer {

    @Throws(SigningException::class)
    override fun sign(message: ByteArray): ByteArray =
        try {
            val prehashed = IrohaHash.prehash(message)
            NativeSignerBridge.signDetached(
                SigningAlgorithm.ML_DSA,
                privateKey.encoded,
                prehashed,
            )
        } catch (ex: RuntimeException) {
            throw SigningException("ML-DSA signing failed", ex)
        }

    override fun publicKey(): ByteArray = keyPublicKey.encoded

    override fun algorithm(): String = SigningAlgorithm.ML_DSA.providerName
}
